datafusion/common/src/tree_node.rs

```diff
@@ -507,37 +495,67 @@ impl<T> Transformed<T> {
                 });
             }
             TreeNodeRecursion::Jump => {
-                self.tnr = return_if_jump;
+                self.tnr = TreeNodeRecursion::Continue;
             }
             TreeNodeRecursion::Stop => {}
         }
         Ok(self)
     }
 
-    /// If [`TreeNodeRecursion`] of the node is [`TreeNodeRecursion::Continue`] or
-    /// [`TreeNodeRecursion::Jump`], transformation is applied to the node.
-    /// Otherwise, it remains as it is.
-    pub fn try_transform_node<F: FnOnce(T) -> Result<Transformed<T>>>(
+    /// Maps the [`Transformed`] object to the result of the given `f` depending on the
+    /// current [`TreeNodeRecursion`] value and the fact that `f` is changing the current
+    /// node's sibling.
+    pub fn transform_sibling<F: FnOnce(T) -> Result<Transformed<T>>>(
         self,
         f: F,
     ) -> Result<Transformed<T>> {
-        if self.tnr == TreeNodeRecursion::Stop {
-            Ok(self)
-        } else {
-            f(self.data).map(|mut t| {
+        match self.tnr {
+            TreeNodeRecursion::Continue | TreeNodeRecursion::Jump => {
+                f(self.data).map(|mut t| {
+                    t.transformed |= self.transformed;
+                    t
+                })
+            }
+            TreeNodeRecursion::Stop => Ok(self),
+        }
+    }
+
+    /// Maps the [`Transformed`] object to the result of the given `f` depending on the
+    /// current [`TreeNodeRecursion`] value and the fact that `f` is changing the current
+    /// node's parent.
+    pub fn transform_parent<F: FnOnce(T) -> Result<Transformed<T>>>(
```

Comment:
likewise here - maybe it would make more sense to call this `transform_self` or `transform_node`